controller/instance_manager_controller.go

```diff
@@ -2044,6 +2087,265 @@ func (m *InstanceManagerMonitor) StopMonitorWithLock() {
+	existReplica, err := m.ds.GetReplicaRO(instanceName)
+	switch {
+	case err == nil:
+		// Replica CR still exists - check the ownership
+		return m.isInstanceScheduledOnNode(&existReplica.Spec.InstanceSpec, &existReplica.Status.InstanceStatus), nil
+	case apierrors.IsNotFound(err):
+		// Replica CR not found - instance is orphaned
+		return false, nil
+	default:
+		// Unexpected error - unable to check if replica instance is orphaned or not
+		return false, err
+	}
+}
+
+// isInstanceScheduledOnNode returns true only when it is very certain that an instance is scheduled on this node
+func (m *InstanceManagerMonitor) isInstanceScheduledOnNode(spec *longhorn.InstanceSpec, status *longhorn.InstanceStatus) bool {
+	if status.CurrentState != spec.DesireState {
+		return false
+	}
+	switch status.CurrentState {
+	case longhorn.InstanceStateRunning:
+		return status.OwnerID == spec.NodeID && spec.NodeID == m.controllerID
+	case longhorn.InstanceStateStopped:
+		return spec.NodeID == m.controllerID
+	default:
+		return false
+	}
+}
+
+func (m *InstanceManagerMonitor) createOrphan(im *longhorn.InstanceManager, instanceName string, orphanType longhorn.OrphanType, dataEngineType longhorn.DataEngineType) error {
+	name := types.GetOrphanChecksumNameForOrphanedInstance(instanceName, m.controllerID, string(dataEngineType))
+
+	if _, err := m.ds.GetOrphanRO(name); err == nil || !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// labels will be attached by mutator webhook
+	orphan := &longhorn.Orphan{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			OwnerReferences: datastore.GetOwnerReferencesForInstanceManager(im),
+		},
+		Spec: longhorn.OrphanSpec{
+			NodeID:     m.controllerID,
+			Type:       orphanType,
+			DataEngine: dataEngineType,
+			Parameters: map[string]string{
+				longhorn.OrphanInstanceName:    instanceName,
+				longhorn.OrphanInstanceManager: im.Name,
+			},
+		},
+	}
+
+	_, err := m.ds.CreateOrphan(orphan)
+	return err
+}
+
+func (m *InstanceManagerMonitor) cleanupOrphans(existOrphans map[string]*longhorn.Orphan, deleteOrphans map[string]bool, autoDeleteEngine, autoDeleteReplica bool) {
+	for orphanName, orphan := range existOrphans {
+		instanceName := orphan.Spec.Parameters[longhorn.OrphanInstanceName]
+		switch orphan.Spec.Type {
+		case longhorn.OrphanTypeEngineInstance:
+			if err := m.cleanupOrphan(orphanName, deleteOrphans, autoDeleteEngine); err != nil {
+				m.logger.WithError(err).WithField("instance", instanceName).Errorf("Failed to cleanup engine orphan %s", orphanName)
+			}
+		case longhorn.OrphanTypeReplicaInstance:
+			if err := m.cleanupOrphan(orphanName, deleteOrphans, autoDeleteReplica); err != nil {
+				m.logger.WithError(err).WithField("instance", instanceName).Errorf("Failed to cleanup replica orphan %s", orphanName)
+			}
+		}
+	}
+}
+
+func (m *InstanceManagerMonitor) cleanupOrphan(orphanName string, deleteOrphans map[string]bool, autoDelete bool) error {
```

Comment:
nit: why not `deleteOrphan`?
```suggestion
func (m *InstanceManagerMonitor) deleteOrphan(orphanName string, deleteOrphans map[string]bool, autoDelete bool) error {
```